File: function_calculator/rational.cpp
/*  Copyright Paul Cernea, May 2025.
All Rights Reserved.*/

#include "rational.h"

#include <stdexcept>
#include <sstream>

namespace FunctionalCalculator
{
  bool parenthesesWellFormed(const std::string& str, const std::pair<char, char>& leftRight)
  {
    auto& left = leftRight.first;
    auto& right = leftRight.second;
    int stackHeight = 0;
    for (const auto& current : str)
    {
      if (current == left) { ++stackHeight; continue; }
      if (current == right)
      {
        if (stackHeight == 0) { return false; }
        --stackHeight;
        continue;
      }
    }
    return (stackHeight == 0);
  }

  Rational::Rational(int nn, int dd)
  {
    *this = Rational(mp(nn), mp(dd));
  }

  Rational::Rational(const mp& nn, const mp& dd)
  {
    if (dd == mp(0))
    {
      throw std::invalid_argument("Division by zero.");
    }
    else if (nn == mp(0))
    {
      num = nn;
      denom = mp(1);
    }
    else
    {
      mp gcd_ = mp::gcd(nn, dd);
      num = nn; denom = dd;
      if (gcd_ != 0)
      {
        num = num / gcd_;
        denom = denom / gcd_;
      }
      if (denom < 0)
      {
        num = num * mp(-1);
        denom = denom * mp(-1);
      }
    }
  }

  Rational::Rational(const Rational& rhs)
  {
    num = rhs.num;
    denom = rhs.denom;
  }

  Rational::Rational(Rational&& rhs) noexcept
  {
    num = rhs.num;
    denom = rhs.denom;
  }

  Rational& Rational::operator=(const Rational& rhs)
  {
    if (this != &rhs)
    {
      num = rhs.num;
      denom = rhs.denom;
    }
    return *this;
  }

  Rational& Rational::operator=(Rational&& rhs) noexcept
  {
    if (this != &rhs)
    {
      num = rhs.num;
      denom = rhs.denom;
    }
    return *this;
  }

  mp Rational::denominator() const { return denom; }
  mp Rational::numerator() const { return num; }

  std::pair<Rational, Rational> Rational::separateSquaredPart() const
  {
    std::pair<Rational, Rational> answer;
    auto sqPartNum = num.separateSquaredPart();
    auto sqPartDen = denom.separateSquaredPart();
    answer.first = Rational(sqPartNum.first, sqPartDen.first);
    answer.first = Rational(sqPartNum.second, sqPartDen.second);
    return answer;
  }

  Rational Rational::operator+() const
  {
    return *this;
  }

  Rational Rational::operator-() const
  {
    return Rational(-num, denom);
  }

  Rational Rational::operator+(const Rational& rhs) const
  {
    return Rational(num * rhs.denom + rhs.num * denom, denom * rhs.denom);
  }

  Rational Rational::operator-(const Rational& rhs) const
  {
    return ((*this) + (-rhs));
  }

  Rational Rational::operator*(const Rational& rhs) const
  {
    return Rational(num * rhs.num, denom * rhs.denom);
  }

  Rational Rational::operator/(const Rational& rhs) const
  {
    if (rhs.num == 0)
    {
      throw std::invalid_argument("Operator division by zero.");
    }
    return Rational(num * rhs.denom, denom * rhs.num);
  }

  Rational Rational::pow(int p) const
  {
    bool isNeg = (p < 0);
    if (isNeg) { p = -p; }
    Rational answer(1, 1);
    for (int i = 0; i < p; ++i)
    {
      answer = answer * (*this);
    }
    if (isNeg)
    {
      return Rational(answer.denom, answer.num);
    }
    return answer;
  }

  bool Rational::operator==(const Rational& rhs) const
  {
    if (rhs.num != num) { return false; }
    if (rhs.denom != denom) { return false; }
    return true;
  }

  bool Rational::operator!=(const Rational& rhs) const
  {
    if (*this == rhs) { return false; }
    return true;
  }

  bool Rational::operator<(const Rational& rhs) const
  {
    if (num * rhs.denom < denom * rhs.num) { return true; }
    return false;
  }

  bool Rational::operator>(const Rational& rhs) const
  {
    if (num * rhs.denom > denom * rhs.num) { return true; }
    return false;
  }

  bool Rational::operator<=(const Rational& rhs) const
  {
    if (num * rhs.denom <= denom * rhs.num) { return true; }
    return false;
  }

  bool Rational::operator>=(const Rational& rhs) const
  {
    if (num * rhs.denom >= denom * rhs.num) { return true; }
    return false;
  }

  std::pair<double, double> Rational::get() const
  {
    double nn = (double)(num.toInt());
    double dd = (double)(denom.toInt());
    return { nn / dd, 0.0 };
  }

  bool Rational::isInt() const
  {
    return (denom == mp(1));
  }

  std::string Rational::print(bool useParentheses) const
  {
    std::stringstream strm;
    if (useParentheses) { strm << "("; }
    if (num == 0) { strm << num; }
    else if (denom == 1) { strm << num; }
    else { strm << num << " / " << denom; }
    if (useParentheses) { strm << ")"; }
    return strm.str();
  }

  std::map<mp, int> Rational::primeFactorization() const
  {
    auto numFactors = num.primeFactorization();
    auto denomFactors = denom.primeFactorization();
    for (auto& iter : denomFactors)
    {
      if (iter.first == 1) { continue; }
      if (numFactors.find(iter.first) == numFactors.end())
      {
        numFactors[iter.first] = -iter.second;
        continue;
      }
      numFactors[iter.first] -= iter.second;
    }
    std::map<mp, int> answer;
    int countFactors = (int)numFactors.size();
    for (auto& iter : numFactors)
    {
      mp base = iter.first;
      int power = iter.second;
      if (base == -1)
      {
        if (power < 0) { power = -power; }
        power = (power % 2);
        if ((countFactors == 1) && (power == 0)) { base = 1; power = 1; }
      }
      if ((base == 1) && (countFactors > 0)) { power = 0; }
      if (power == 0)
      {
        if (countFactors == 1) { base = 1; power = 1; }
        else { continue; }
      }
      answer[base] = power;
    }
    return answer;
  }

  std::string Rational::printFactors(bool useParentheses) const
  {
    std::stringstream strm;
    if (useParentheses) { strm << "("; }
    auto factors = primeFactorization();
    int count = -1;
    for (auto& iter : factors)
    {
      if (iter.second == 0) { continue; }
      ++count;
      if (count > 0) { strm << " * "; }
      if (iter.first < 0) { strm << "("; }
      strm << iter.first;
      if (iter.first < 0) { strm << ")"; }
      strm << "^" << iter.second;
    }
    if (useParentheses) { strm << ")"; }
    return strm.str();
  }
}

File: function_calculator_cs/rational.cs
/*  Copyright Paul Cernea, June 2025.
All Rights Reserved.*/

namespace function_calculator_cs
{

public class Rational
{
  private mp num = new mp(0);
  private mp denom = new mp(1);

  static Rational()
  {
  }

  public Rational(int nn = 0, int dd = 1) : this(new mp(nn), new mp(dd))
  {
  }

  public Rational(in mp nn, in mp dd)
  {
    var zero_ = mp.zero();
    if (dd == zero_)
    {
      throw new System.Exception("Division by zero.");
    }
    else if (nn == zero_)
    {
      num = new mp(nn);
      denom = new mp(1);
    }
    else
    {
      mp gcd_ = mp.gcd(nn, dd);
      num = new mp(nn); denom = new mp(dd);
      if (gcd_ != zero_)
      {
        num = num / gcd_;
        denom = denom / gcd_;
      }
      if (denom < zero_)
      {
        num = num * (new mp(-1));
        denom = denom * (new mp(-1));
      }
    }
  }

  public Rational(in Rational rhs)
  {
    if (rhs is null) { throw new System.Exception("Copy constructor given null parameter."); }
    num = rhs.num;
    denom = rhs.denom;
  }

  public static Rational zero() { return new Rational(0, 1); }

  public mp denominator() { return new mp(denom); }
  public mp numerator() { return new mp(num); }

  /** \return { a, b } where a is the maximal number such that this number == a * a * b */
  public KeyValuePair<Rational, Rational> separateSquaredPart()
  {
    var sqPartNum = num.separateSquaredPart();
    var sqPartDen = denom.separateSquaredPart();
    KeyValuePair<Rational, Rational> answer = new KeyValuePair<Rational, Rational>(
      new Rational(sqPartNum.Key, sqPartDen.Key),
      new Rational(sqPartNum.Value, sqPartDen.Value));
    return answer;
  }

  public static Rational operator+(in Rational body) { return new Rational(body); }
  public static Rational operator-(in Rational body)
  {
    return new Rational(-body.num, body.denom);
  }

  public static Rational operator+(in Rational body, in Rational rhs)
  {
    return new Rational(body.num * rhs.denom + rhs.num * body.denom, body.denom * rhs.denom);
  }

  public static Rational operator-(in Rational body, in Rational rhs)
  {
    return (body + (-rhs));
  }

  public static Rational operator*(in Rational body, in Rational rhs)
  {
    return new Rational(body.num * rhs.num, body.denom * rhs.denom);
  }

  public static Rational operator/(in Rational body, in Rational rhs)
  {
    if (rhs.num == mp.zero())
    {
      throw new System.Exception("Operator division by zero.");
    }
    return new Rational(rhs.num * rhs.denom, rhs.denom * rhs.num);
  }

  public Rational pow(int p) /**< Returns the p'th power of the rational number. */
  {
    Boolean isNeg = (p < 0);
    if (isNeg) { p = -p; }
    Rational answer = new Rational(1, 1);
    for (int i = 0; i < p; ++i)
    {
      answer = answer * (this);
    }
    if (isNeg)
    {
      return new Rational(answer.denom, answer.num);
    }
    return answer;
  }

  public static Boolean operator==(in Rational body, in Rational rhs) { return body.Equals(rhs); }

  public static Boolean operator!=(in Rational body, in Rational rhs)
  {
    if (body == rhs) { return false; }
    return true;
  }

  public static Boolean operator<(in Rational body, in Rational rhs)
  {
    if (body.num * rhs.denom < body.denom * rhs.num) { return true; } // denom always > 0
    return false;
  }

  public static Boolean operator>(in Rational body, in Rational rhs) { return (rhs < body); }
  public static Boolean operator<=(in Rational body, in Rational rhs)
  {
    if (body == rhs) { return true; }
    return (body < rhs);
  }

  public static Boolean operator>=(in Rational body, in Rational rhs)
  {
    if (body == rhs) { return true; }
    return (body > rhs);
  }

  /** \brief The keys are the prime factors, the values are the number of occurrences. */
  public Dictionary<mp, int> primeFactorization()
  {
    var numFactors = num.primeFactorization();
    var denomFactors = denom.primeFactorization();
    var one_ = new mp(1);
    foreach (var iter in denomFactors)
    {
      if (iter.Key == one_) { continue; }
      int iterSecond = 0;
      Boolean found = numFactors.TryGetValue(iter.Key, out iterSecond);
      if (!found)
      {
        numFactors[iter.Key] = -iter.Value;
        continue;
      }
      numFactors[iter.Key] -= iter.Value;
    }
    Dictionary<mp, int> answer = new Dictionary<mp, int>();
    int countFactors = (int)numFactors.Count;
    foreach (var iter in numFactors)
    {
      mp baseMp = new mp(iter.Key);
      int power = iter.Value;
      if (baseMp == new mp(-1))
      {
        if (power < 0) { power = -power; }
        power = (power % 2);
        if ((countFactors == 1) && (power == 0)) { baseMp = new mp(one_); power = 1; }
      }
      if ((baseMp == one_) && (countFactors > 0)) { power = 0; }
      if (power == 0)
      {
        if (countFactors == 1) { baseMp = new mp(one_); power = 1; }
        else { continue; }
      }
      answer[baseMp] = power;
    }
    return answer;
  }

  /** \brief Print the prime factorization of the rational number. */
  public string printFactors(Boolean useParentheses = false)
  {
    string strm = "";
    var zero_ = mp.zero();
    if (useParentheses) { strm += "("; }
    var factors = primeFactorization();
    int count = -1;
    foreach (var iter in factors)
    {
      if (iter.Value == 0) { continue; }
      ++count;
      if (count > 0) { strm += " * "; }
      if (iter.Key < zero_) { strm += "("; }
      strm += iter.Key;
      if (iter.Key < zero_) { strm += ")"; }
      strm += "^" + iter.Value;
    }
    if (useParentheses) { strm += ")"; }
    return strm;
  }

  public virtual double toDouble()
  {
    double nn = (double)(num.toInt());
    double dd = (double)(denom.toInt());
    return nn / dd;
  }

  public Boolean isInt()
  {
    return (denom == new mp(1));
  }

  public override string ToString()
  {
    return ToString(false);
  }

  public string ToString(Boolean useParentheses)
  {
    string strm = "";
    if (num == mp.zero()) { strm = strm + num; }
    else if (denom == new mp(1)) { strm = strm + num; }
    else { strm  = strm + num + " / " + denom; }
    if (useParentheses) { return "(" + strm + ")"; }
    return strm;
  }

  public override bool Equals(object? obj)
  {
    return Equals(obj as Rational);
  }

  public bool Equals(Rational? other)
  {
    if (other is null) { return false; }
    if (other.num != num) { return false; }
    if (other.denom != denom) { return false; }
    return true;
  }

  public override int GetHashCode()
  {
    return HashCode.Combine(num, denom);
  }
}
}

